### DISCUSSION TITLE
gba.js save states

### DISCUSSION THREAD

**Discussion by city41** [macabeus/klo-gba.js#94] [Show and tell]
Hi! Thanks for making this project and the Medium posts you wrote about it. It's been very helpful for me, as I'm making a similar tool for Super Mario Advance 4.

I just wanted to let you know I just got save states working in gba.js, so now in my Mario editor, the user can jump into their level really fast. I juuuuuust got them working, so the code is pretty bad. But if you want to add save state support to klo-gba, should be able to do it.

My project is here: https://github.com/city41/smaghetti

**Comment by macabeus** [comment]
Very thank you! It was one of the missing features that I really wanted to add.
And very nice to hear that my project and posts was useful for your project =]

Before putting this project on hold, I was looking to extract `react-gbajs` into a new package. Maybe it would be useful to do now?

**Reply by city41** [reply]
I like the idea of react-gbajs. I have also been toying with the idea of forking gbajs and improving it. I was thinking add save state support, port it to typescript with ES modules, improve the performance in firefox (or at least, try, might not be possible), finish the webgl renderer, etc

I'm on the fence though, as I don't want to take on a bunch of gbajs work that would prevent me from working on my editor. Not sure what the right balance is.

**Reply by macabeus** [reply]
I don't know how much would be useful to spend much time improving `gbajs` itself, since its author said that is better to avoid to use it since they is launching a webassembly version for mgba: https://github.com/mgba-emu/mgba/issues/1296

`react-gbajs` is using `gbajs` just as a workaround while mgba webasembly isn't finished.

And for the long term, IMHO is better to use the webassembly version into `react-gbajs` replacing the `gbajs`.

**Reply by city41** [reply]
I actually don't plan to switch to mgba. Web assembly is great, but I prefer to avoid it if it's not needed. And honestly, gbajs works great, I find it's full speed in all browsers except Firefox. But who knows, maybe the Firefox perf issues will cause me to switch mgba.

**Reply by city41** [reply]
Anyway, as for save states, maybe the quickest approach is to do react-gbajs and add save state support there?

**Reply by city41** [reply]
Oh yeah, I also forgot, for save states I haven't figured out sound. After doing a restore, sound does not work. But I don't think that will be hard to fix. I think it will be similar to how the fix for video was, basically just need to dig in and find all sound related state and extract it.

**Reply by macabeus** [reply]
Yeah. I like this idea too.

In klo-gba.js, [react-gbajs](https://github.com/macabeus/klo-gba.js/tree/master/react-gbajs) is almost independent of any other project here.
If you want, you can extract it from this project and create a new one, or making a PR here and I create an NPM package later.

I'll have time this weekend to help you with this task.

**Reply by city41** [reply]
Ok cool. I think I'll fork klo-gba and do a PR, and just keep react-gbajs where it is for now. 

I'll see about getting this going this weekend.

**Reply by city41** [reply]
Hmmm, when I try to get klo-gba running on my machine I get

```
scissors.js?:formatted:42674 Uncaught RuntimeError: abort(You are setting onRuntimeInitialized on the Promise object, instead of the instance. Use .then() to get called back with the instance, see the MODULARIZE docs in src/settings.js) at Error
    at jsStackTrace (webpack:///../scissors/dist/scissors.js?:8979:494)
    at stackTrace (webpack:///../scissors/dist/scissors.js?:8981:150)
    at abort (webpack:///../scissors/dist/scissors.js?:8943:190)
    at Promise.set (webpack:///../scissors/dist/scissors.js?:8627:6632)
```

I followed the readme's setup instructions and built the wasm successfully and ran `yarn start` in each of the projects.

**Reply by macabeus** [reply]
It's weird... I did clone and setup up the project again, and I got to reproduce this bug. I really didn't notice when it started to happen.
I'm trying to fix it now.

**Reply by macabeus** [reply]
@city41 I'm still investigating more about the bug, but just to unblock you: apply the following patch:
```diff
diff --git a/scissors/src/huffman.js b/scissors/src/huffman.js
index da87d49..f94f03b 100644
--- a/scissors/src/huffman.js
+++ b/scissors/src/huffman.js
@@ -19,7 +19,7 @@ class HuffmanDecodeError extends Error {
   }
 }
 
-huffmanModule.onRuntimeInitialized = () => {}
+// huffmanModule.onRuntimeInitialized = () => {}
 
 const huffmanDecode = (buffer) => {
   FS.writeFile('file', buffer)
diff --git a/scissors/src/lzss.js b/scissors/src/lzss.js
index a126bbb..8337b27 100644
--- a/scissors/src/lzss.js
+++ b/scissors/src/lzss.js
@@ -19,7 +19,7 @@ class LzssDecodeError extends Error {
   }
 }
 
-lzssModule.onRuntimeInitialized = () => {}
+// lzssModule.onRuntimeInitialized = () => {}
 
 const lzssDecode = (buffer) => {
   FS.writeFile('filelzss', buffer)
``` 

Probably it'll fix the bug.

**Reply by macabeus** [reply]
I forgot to set a specific version on the docker image. Probably it's the root of the bug 🤦
https://github.com/macabeus/klo-gba.js/blob/b1f5cd9ed326b7f2850a1d2f764345b9d9dad9df/shellscript/compile-webassembly.sh#L9

**Reply by macabeus** [reply]
Done. Fixed: https://github.com/macabeus/klo-gba.js/pull/95

**Reply by city41** [reply]
awesome! thanks for the quick fix!
